12/umbraco-cms/reference/routing/custom-middleware.md

```diff
@@ -0,0 +1,70 @@
+- `WithCustomMiddleware()` is a method that can be used in Umbraco for adding custom middleware. This includes some specific customizable instructions that run in the request processing pipeline. This method is often used in combination with IUmbracoPipelineFilter.
+
+{% hint style="warning" %}
+
+`WithCustomMiddleware` should only be used as a last resort, as Umbraco can break if you forget to add middleware or add them in the wrong order.  
+
+{% endhint %}
+
+## Configuring the Cross-Origin Resource Sharing (CORS) middleware
+
+Create a composer with the following:
+
+```csharp
+
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Web.Common.ApplicationBuilder;
+
+public class CorsComposer : IComposer
+{
+    public const string AllowAnyOriginPolicyName = nameof(AllowAnyOriginPolicyName);
+
+    public void Compose(IUmbracoBuilder builder)
+        => builder.Services
+        .AddCors(options => options.AddPolicy(AllowAnyOriginPolicyName, policy => policy.AllowAnyOrigin()))
+        .Configure<UmbracoPipelineOptions>(options => options.AddFilter(new UmbracoPipelineFilter("Cors", postRouting: app => app.UseCors())))
+        // For testing only
+        .Configure<UmbracoPipelineOptions>(options => options.AddFilter(new UmbracoPipelineFilter("CorsTest", endpoints: app => app.UseEndpoints(endpoints =>
+        {
+            endpoints.MapGet("/echo", context => context.Response.WriteAsync("echo")).RequireCors(AllowAnyOriginPolicyName);
+            endpoints.MapGet("/echo2", context => context.Response.WriteAsync("echo2"));
+        }))));
+}
+```
+
+You should be able to request /echo from any origin, but get an error when trying to fetch /echo2. You can test this by pasting the following JavaScript code in your browser console when not on the local Umbraco website (such as Umbraco.com):
+
+```javascript
+fetch('https://localhost:44331/echo', {method: 'GET'}).then(result => result.text().then(text => console.log(text)));
+fetch('https://localhost:44331/echo2', {method: 'GET'}).then(result => result.text().then(text => console.log(text)));
+```
+
+This should return the following:
+<figure><img src="../images/custom-middleware-cors-browser-example.png" alt=""><figcaption><p>Browser result image</p></figcaption></figure>
+
+An additional bonus is that this doesn't require any changes in your `Startup.cs` file. It also allows packages to enable Cross-Origin Resource Sharing (CORS) and configure their own policies.
+
+Users that currently use `WithCustomMiddleware()` will need to add calls to `RunPreRouting()` and `RunPostRouting()`. This is similar to Umbraco adding additional middleware in future versions.
+
+You can find more information about this example in the github PR [Add PreRouting and PostRouting pipeline filters](https://github.com/umbraco/Umbraco-CMS/pull/14503)
```

Comment:
Do we actually refer to PRs in our documentation or should we make sure everything is covered in this page?
```suggestion
You can find more information about this example in the GitHub PR [Add PreRouting and PostRouting pipeline filters](https://github.com/umbraco/Umbraco-CMS/pull/14503).
```